internal/backend/runtime/omni/controllers/omni/internal/exposedservice/reconciler.go

```diff
@@ -0,0 +1,300 @@
+		}
+	}
+
+	return servicesToKeep, nil
+}
+
+func (reconciler *Reconciler) reconcileService(ctx context.Context, exposedService *omni.ExposedService, service *corev1.Service, logger *zap.Logger) (keep bool, err error) {
+	existingExposedService := reconciler.exposedServices[exposedService.Metadata().ID()]
+	if existingExposedService != nil && existingExposedService.Metadata().Phase() == resource.PhaseTearingDown {
+		return false, nil
+	}
+
+	port, err := strconv.Atoi(service.Annotations[ServicePortAnnotationKey])
+	if err != nil || port < 1 || port > 65535 {
+		logger.Warn("invalid port on Service", zap.String("port", service.Annotations[ServicePortAnnotationKey]))
+
+		return true, nil //nolint:nilerr
+	}
+
+	label, labelOk := service.Annotations[ServiceLabelAnnotationKey]
+	if !labelOk {
+		label = service.Name + "." + service.Namespace
+	}
+
+	icon, err := reconciler.parseIcon(service.Annotations[ServiceIconAnnotationKey])
+	if err != nil {
+		logger.Debug("invalid icon on Service", zap.Error(err))
+	}
+
+	explicitAliasOpt := optional.None[string]()
+	if explicitAlias, ok := service.Annotations[ServicePrefixAnnotationKey]; ok {
+		explicitAliasOpt = optional.Some(explicitAlias)
+	}
+
+	updatedService, err := safe.WriterModifyWithResult(ctx, reconciler.r, exposedService, func(res *omni.ExposedService) error {
+		return reconciler.updateExposedService(res, explicitAliasOpt, port, label, icon, logger)
+	})
+	if err != nil {
+		return false, fmt.Errorf("error updating exposed service: %w", err)
+	}
+
+	alias, _ := updatedService.Metadata().Labels().Get(omni.LabelExposedServiceAlias)
+	reconciler.usedAliases[alias] = updatedService.Metadata().ID()
+	reconciler.exposedServices[updatedService.Metadata().ID()] = updatedService
+
+	return true, nil
+}
+
+func (reconciler *Reconciler) updateExposedService(res *omni.ExposedService, explicitAliasOpt optional.Optional[string], port int, label, icon string, logger *zap.Logger) error {
+	res.Metadata().Labels().Set(omni.LabelCluster, reconciler.cluster)
+
+	explicitAlias, explicitAliasRequested := explicitAliasOpt.Get()
+	_, hasGeneratedAlias := res.Metadata().Labels().Get(omni.LabelHasGeneratedExposedServiceAlias)
+
+	var (
+		alias string
+		err   error
+	)
+
+	switch {
+	case explicitAlias != "": // replace the existing alias with the explicit one
+		alias = strings.ToLower(explicitAlias)
+
+		if currentOwner, ok := reconciler.usedAliases[alias]; ok && currentOwner != res.Metadata().ID() {
+			res.TypedSpec().Value.Error = fmt.Sprintf("requested alias %q is already used by another service", alias)
+
+			logger.Warn(res.TypedSpec().Value.Error)
+
+			return nil
+		}
+	case !hasGeneratedAlias: // no alias, generate one
+		if alias, err = reconciler.generateExposedServiceAlias(); err != nil {
+			return fmt.Errorf("error generating exposed service alias: %w", err)
+		}
+	default: // keep the existing alias
+		alias, _ = res.Metadata().Labels().Get(omni.LabelExposedServiceAlias)
+	}
```

Comment:
This is the actual change in the logic:
- Check if user requested an explicit prefix (set the annotation on the k8s service)
- If they did, use that alias instead of generating a new one
- When using that alias, make sure that it is not already taken. If it is, save the error on the resource. First claimer of the prefix gets to keep it.
- When user opts out from an explicit alias (removes the annotation), figure that out by using the new annotation `omni.LabelHasGeneratedExposedServiceAlias`, and generate a fresh one.